File: algorithms/PopulatingNextRightPointersinEachNode/PopulatingNextRightPointersinEachNode.c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */
void connect(struct TreeLinkNode *root) {
	if (root != NULL && root->left != NULL && root->right != NULL)
	{
		root->left->next = root.right;
		if (root->next)
			root->right->next = root->next->left;
		connect(root->left);
		connect(root->right)
	}
}

File: algorithms/PopulatingNextRightPointersinEachNode/PopulatingNextRightPointersinEachNode.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect1(self, root):
        if root and root.left and root.right:
            root.left.next = root.right
            if root.next:
                root.right.next = root.next.left
            self.connect(root.left)
            self.connect(root.right)

    #BFS
    def connect2(self, root):
        if not root:
            return
        queue = [root]
        while queue:
            curr = queue.pop(0)
            if curr.left and curr.right:
                curr.left.nect = curr.right
                if curr.next:
                    curr.right.next = curr.next.left
                queue.append(curr.left)
                queue.append(curr.right)
